### PR TITLE
api: add option metadata

### PR DESCRIPTION
**Pull request by tjdevries**

### CODE REVIEW

**Comment by tjdevries** [comment]
@teto  I was thinking maybe we could just expose an API function that does what you were thinking of doing with `options.lua` -- since we can just generate whatever we want in our code :)

This might be the easiest way to expose this without adding any new deps / files

**Comment by teto** [comment]
Yes that may be a good idea, avoid having outdated entries. 
Some options name may be tough to find without a description  so I  still like https://github.com/neovim/neovim/pull/6288. Even if the description is not embedded in the binary it could still be useful for 3rd parties.

I think the `nvim_set_smartoption` should be in a different PR (different problem). It tells something about our API that we need such a function xD

**Comment by bfredl** [comment]
progress: `print(vim.inspect(vim.api.nvim_get_options_info()))` works, but `echo nvim_get_options_info()` crashes. Lesson: just use lua already.

File: src/nvim/option.c [hunk 1]
```diff
@@ -7174,3 +7175,57 @@ long get_sidescrolloff_value(void)
...
+{
+    Dictionary dict = ARRAY_DICT_INIT;
+
+    PUT(dict, "name", STRING_OBJ(cstr_to_string(opt->fullname)));
+    PUT(dict, "shortname", STRING_OBJ(cstr_to_string(opt->shortname)));
+    PUT(dict, "global",  BOOLEAN_OBJ(opt->flags & OPT_GLOBAL));
+    PUT(dict, "local",  BOOLEAN_OBJ(opt->flags & OPT_LOCAL));
+    PUT(dict, "flag",  INTEGER_OBJ(opt->flags));  // TODO(bfredl): lol tj
+    PUT(dict, "last_set_sid", INTEGER_OBJ(opt->last_set.script_ctx.sc_sid));
+    PUT(dict, "last_set_linenr", INTEGER_OBJ(opt->last_set.script_ctx.sc_lnum));
```
Comment (bfredl) [inline src/nvim/option.c:7209]: TODO: what about the channel number

**Comment by tjdevries** [comment]
@bfredl I think this is good now? I didn't like the having keys sometimes being in the dictionarty and sometimes not (for languages like Python, where you don't just silenty get returned nil for access a dict). So I changed it so that the keys are ALWAYS in the dictionary.

I don't think we actually need to worry about the description thing I mentioned in gitter. If you think it looks good and the tests pass tonight, you can merge when you wake up.

**Comment by bfredl** [comment]
> I didn't like the having keys sometimes being in the dictionarty and sometimes not (for languages like Python, where you don't just silenty get returned nil for access a dict). So I changed it so that the keys are ALWAYS in the dictionary.

It that case we should merge `win` and `buf` to a single key: `scope` which the value either `win`, `buf` or `global`. ( but leave `global_local` as it is).

**Comment by bfredl** [comment]
Done. all the keys are now always present, but `win` and `buf` was merged to one (to not suggest a hypothetical `win=true, buf=true` option kind)

File: src/nvim/api/vim.c [hunk 2]
```diff
@@ -970,6 +970,47 @@ Object nvim_get_option(String name, Error *err)
   return get_option_from(NULL, SREQ_GLOBAL, name, err);
 }
 
+/// Gets the option information for all options.
+///
+/// The dictionary has the full option names as keys and option metadata
+/// dictionaries as detailed at |nvim_get_option_info|.
+///
+/// @return dictionary of all options
+Dictionary nvim_get_all_options_info(Error *err)
```
Comment (justinmk) [inline src/nvim/api/vim.c:979]: "list" would be the verb suggested by `:help dev-api` , and is consistent with existing API function names

**Pull request merged**